clippy_lints/src/loops.rs

```diff
@@ -293,18 +293,48 @@ declare_clippy_lint! {
 declare_clippy_lint! {
     /// **What it does:** Checks for empty `loop` expressions.
     ///
-    /// **Why is this bad?** Those busy loops burn CPU cycles without doing
-    /// anything. Think of the environment and either block on something or at least
-    /// make the thread sleep for some microseconds.
+    /// **Why is this bad?** Due to [an LLVM codegen bug](https://github.com/rust-lang/rust/issues/28728)
+    /// these expressions can produce undefined behavior. Also, these busy loops
+    /// burn CPU cycles without doing anything.
     ///
-    /// **Known problems:** None.
+    /// It is _almost always_ a better idea to `panic!` than to have a busy loop.
+    ///
+    /// If panicking isn't possible, think of the environment and either:
+    ///   - block on something
+    ///   - sleep the thread for some microseconds
+    ///   - yield or pause the thread
+    ///
+    /// For `std` targets, this can be done with
+    /// [`std::thread::sleep`](https://doc.rust-lang.org/std/thread/fn.sleep.html)
+    /// or [`std::thread::yield_now`](https://doc.rust-lang.org/std/thread/fn.yield_now.html).
+    ///
+    /// For `no_std` targets, doing this is more complicated, especially because
+    /// `#[panic_handler]`s can't panic. To stop/pause the thread, you will
+    /// probably need to invoke some target-specific intrinsic. Examples include:
+    ///   - [`x86_64::instructions::hlt`](https://docs.rs/x86_64/0.12.2/x86_64/instructions/fn.hlt.html)
+    ///   - [`cortex_m::asm::wfi`](https://docs.rs/cortex-m/0.6.3/cortex_m/asm/fn.wfi.html)
+    ///
+    /// If you just care about fixing the LLVM bug (and don't care about burning
+    /// CPU), you can:
+    ///   - On nightly, compile your code with `-Z insert-sideeffect`
```

Comment:
Should we also suggest the `llvm_asm!` version here for nightly? That way the user won't have to set any compiler flags and just has to change some code. Also the `llvm_asm!` version is a bit better than the `read_volatile` version, since it doesn't insert any code. (Even though the inserted code doesn't really make a difference performance wise or in any other way, since, as you stated, the program will be trapped in an infinite loop)

Reply:
Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

`-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

Reply:
> Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

The old `asm!` got renamed to `llvm_asm!`. A new `asm!` was added (which is way better) that has a path to stablization.

> `-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

It becoming the default behavior is actually my concern (as then the flag would be removed).

Reply:
https://github.com/rust-lang/rfcs/pull/2843 / https://github.com/rust-lang/rust/issues/70173

Ah `asm!` should be reimplemented, so it can become stable sometime in the future. Alright, in that case we should suggest to use `asm!`

Reply:
> Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

Indeed. `llvm_asm!` is the new thing.

> `-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

The `-Z` flag is most certainly perma unstable, it may resurface under a `-C` flag or become enabled by default (which I really do not hope because it's just a bad and costly workaround), so whatever the stable long term solution will be, it's certainly not a `-Z` flag.

Reply:
> The `-Z` flag is most certainly perma unstable, it may resurface under a `-C` flag

Oh yeah, `-Z` flags are of course perma unstable, oops. Time to sleep 🛌

Reply:
I changed the recommendation to use an empty assembly statement `asm!("")`. It isn't marked `pure`, so this therefore a sideeffect.